API.md

```diff
@@ -0,0 +1,329 @@
+- the way in which OpenSCD sets CSS fonts and colors for plugins
+
+## Communicating data to plugins
+
+OpenSCD core communicates the data necessary for editing SCL documents by setting the following [properties](https://developer.mozilla.org/en-US/docs/Glossary/Property/JavaScript) on the plugin's [DOM Element](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement):
+
+
+```typescript
+export default class MyPlugin extends LitElement {
+  docs: Record<string, XMLDocument> = {};
+  doc?: XMLDocument;
+  docName?: string;
+  editable: string[] = [
+    'cid',
+    'icd',
+    'iid',
+    'scd',
+    'sed',
+    'ssd',
+  ];
+  history: LogEntry[];
+  editCount: number = 0;
+  plugins: { menu: Plugin[], editor: Plugin[] }[];
+  locale: string = 'en';
+}
+
+/** Helper types exported by OpenSCD core **/
+
+type LogEntry = {
+  redo: Edit; // `Edit` defined below
+  undo: Edit;
+  title?: string
+}
+
+type Plugin = {
+  name: string;
+  translations?: Record<string, string>;
+  src: string;
+  icon: string; // Material icon name or image URL
+  requireDoc?: boolean; // disable plugin if no doc is opened
+}
+```
+
+### `docs`
+
+`docs` is set to an object mapping `string` keys (document names) to `XMLDocument` values.
+
+### `docName`
+The name of the `XMLDocument` currently being edited.
+
+### `doc`
+The `XMLDocument` currently being edited.
+
+### `editable`
+Filename extensions of user-editable documents.
+
+### `history`
+History of edits done to `doc`.
+
+### `editCount`
+
+Index of the current log entry in the `history`. Is incremented with every new edit and updated on undo/redo. 
+
+### `plugins`
+
+Arrays of `Plugin` objects describing the currently loaded `menu` and `editor` plugins, respectively.
+
+### `locale`
+
+Selected language (IETF language tag).
+
+## Communicating user intent to OpenSCD core
+
+Plugins communicate user intent to OpenSCD core by dispatching the following [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent):
+
+### `EditEventV2`
+
+The **edit event** allows a plugin to describe the changes it wants to make to the current `doc`.
+
+```typescript
+export type EditDetailV2<E extends Edit = Edit> = {
+  edit: E;
+  title?: string;
+  squash?: boolean;
+}
+
+export type EditEventV2<E extends Edit = Edit> = CustomEvent<EditDetailV2<E>>;
+
+export type EditEventOptions = {
+  title?: string;
+  squash?: boolean;
+}
+
+export function newEditEventV2<E extends Edit>(edit: E, options: EditEventOptions): EditEventV2<E> {
+  return new CustomEvent<E>('oscd-edit-v2', {
+    composed: true,
+    bubbles: true,
+    detail: {...options, edit},
+  });
+}
+
+declare global {
+  interface ElementEventMap {
+    ['oscd-edit-v2']: EditEventV2;
+  }
+}
+```
+
+Its `title` property is a human-readable description of the edit for displaying in the editing history dialog.
+
+The `squash` flag indicates whether the edit should be merged with the previous edit in the history.
```

Comment:
I am not sure if I get this correctly so here is an example:
```ts
const edit0 = { title: "some meaningful message", squash: false, ... }
// from this point forward, I don't need to a title because it is going to be squashed with the previous one
const edit1 = { title: undefined, squash: true, ... }
const edit1 = { title: undefined, squash: true, ... }
```

what happens if I send `squash:true` and so smuggle my change into a history item that is not mine?

```ts
const editFromPluginA = { title:"some uncritical change", ... }

const editFromPluginB = { title:"I will be part of the previous history item", squash:true }
```

another edge case would be if I send the first change with `squash:true` and there is no item to squash with. I assume we should just silently not do anything.


My first thought was that if I want to group changes into a single history item we could send an array of changes with a title. Something along the lines of:

```ts
const batchChanges = {
  title:"I've made a lots of changes",
  squash: true.
  events: [
   { ... },
   { ... },
   { ... },
  ]
}
```
In this case `squash` would be also redundant maybe.

However as I understand some event have dependencies on each other. e.g.:

```ts
const addSomeItemEvent = { title:"added flux capacitor", squash: false }
const initFluxCapacitorCapacitiy = { title: undefined, squash: true }
```
so I (maybe) cannot batch these together because the element in the second event does not exist and I have to wait for it.
Do I get this right?


The concept of "squashing with previous" feels somewhat error-prone. Would be "squash with this ID" too much?
e.g:
```ts
const aRandomId = generateID()
const addSomeItemEvent = { title:"added flux capacitor", id: aRandomId }
const initFluxCapacitorCapacitiy = { title: undefined, squashWith: aRandomId }
```

Reply:
"squash with ID" would actually turn out more problematic in terms of edge-cases, our undo/redo history really has to be completely linear and in sync with the clock time at which the edit is composed. This sad circumstance is directly due to us using the *mutable* `DOM` API for data exchange, referencing live `Node`s and `Element`s in our `Edit`s.

The two edge cases you mention are straightforward to resolve:
- If the `history` is empty, a new `Edit` is created as if `squash` was `false`.
- If the last `Edit` was dispatched by a different plugin, we still allow `squash`ing my own edit onto it.
  This allows for a real improvement of the user experience in certain scenarios. @danyill 's manufacturer-specific "automatic supervision" plugins, for example, might prefer `squash`ing onto the previous edit, their edits thereby being un-done and re-done at the same time as the subscription they are supervising. This would also enable completely new use-cases like "automatic reference update" or "automatic cleanup" plugins that remove orphaned nodes along with the edits that orphaned them or update references automatically along with entity renaming or moving edits, thereby resulting in a valid document at each point along the undo/redo history.

Luckily, all JavaScript runs on a single thred, with all events being processed in order. So as long as a plugin dispatches a series of edits without explicitly calling or yielding to other code in between, it can completely count on the document not being changed from under it in between edits. JavaScript functions thereby behave as transactions when it comes to our `Edit` API.

Given these thoughts, do you agree that the current solution of "squash with previous" is good enough for the use cases that need it? If so, do you still see other use cases that require a more involved API to solve?